adoc/extensions/sycl_khr_command_graph.adoc

```diff
@@ -252,111 +253,423 @@ a@ Constructs a [code]#depends_on_all_leaves# property instance.
+<<subsec:khr-graph-node-properties>>.
+
+_Returns:_ The empty node which has been added to the graph.
+
+_Throws:_ An [code]#exception# with the [code]#errc::invalid# error code if a
+queue is recording commands to the graph.
+
+.[apititle]#khr::command_graph::add#
+[source, role=synopsis,id=api:khr-command-graph-add]
+----
+template <typename T>
+  node add(T cgf, const property_list& propList = {})
+----
+_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.
+
+_Effects:_ The [code]#cgf# command group function behaves in much the same way
+as the command group function passed to [code]#queue::submit# unless explicitly
+stated otherwise.
+Code in the function is evaluated synchronously, before the function returns
+back to [code]#command_graph::add#, with the exception of any SYCL commands
+(e.g. kernels, or explicit memory copy operations).
+These commands are captured into the graph and executed asynchronously when the
+graph is submitted to a queue.
+The requisites of [code]#cgf# will be used to identify any dependent nodes in
+the graph to form edges with.
+Zero or more properties can be provided to this function via an instance of
+[code]#property_list#.
+Valid properties for this function are defined in
+<<subsec:khr-graph-node-properties>>.
+
+
+_Returns:_ The node which has been added to the graph.
+
+_Throws:_
+
+* An [code]#exception# with the [code]#errc::invalid# error code if a queue is
+  recording commands to the graph.
+* An [code]#exception# with the [code]#errc::invalid# error code if the graph
+  wasn't created with the
+  [code]#property::command_graph::assume_buffer_outlives_graph# property and
+  this command uses a buffer.
+ 
+.[apititle]#khr::command_graph::make_edge#
+[source, role=synopsis,id=api:khr-command-graph-make-edge]
 ----
+void make_edge(node& src, node& dest)
+----
+_Constraints:_ [code]#State# is [code]#graph_state::modifiable#.
+
+_Effects:_ Creates a dependency between two nodes representing a happens-before
+relationship.
+Node [code]#dest# will be dependent on [code]#src#.
+
+_Throws:_
+
+* An [code]#exception# with the [code]#errc::invalid# error code if a queue is
+  recording commands to the graph object.
 
-==== Graph Member Functions
+* An [code]#exception# with the [code]#errc::invalid# error code if [code]#src#
+  or [code]#dest# are not valid nodes assigned to the graph object.
 
-===== Constructor of the `command_graph` class
+* An [code]#exception# with the [code]#errc::invalid# error code if [code]#src#
+  and [code]#dest# are the same node.
+
+* An [code]#exception# with the [code]#errc::invalid# error code if the
+  resulting dependency would lead to a cycle in the graph.
+  This error is omitted when [code]#property::command_graph::no_cycle_check# is
+  set.
+
+.[apititle]#khr::command_graph::print_graph#
+[source, role=synopsis,id=api:khr-command-graph-print-graph]
+----
+void print_graph(std::string path, bool verbose = false) const
```

Comment:
Having SYCL interface with file system paths could be complicated, maybe we should consider instead having `print_graph` instead print to a std::string, so then the user can decide what to do with it.

Reply:
That's a fair point, but to use the tools that work on the dot language you probably want a file, and right now this feature is very useful. It's really easy to add to your application for debugging without having to write it to a file yourself. 

We could have 2 versions of this API one that write to a std::string if the current API throws an exception because it can't write to a file?